modules/rostests/kmtests/ntos_ob/ObInsert.c

```diff
@@ -0,0 +1,651 @@
+           Object,
+           ProcessHandleCount,
+           SystemHandleCount,
+           GrantedAccess);
+}
+
+static
+VOID
+NTAPI
+DeleteProc(IN PVOID Object)
+{
+    DPRINT("DeleteProc() 0x%p\n", Object);
+}
+
+static
+NTSTATUS
+NTAPI
+ParseProc(
+    IN PVOID ParseObject,
+    IN PVOID ObjectType,
+    IN OUT PACCESS_STATE AccessState,
+    IN KPROCESSOR_MODE AccessMode,
+    IN ULONG Attributes,
+    IN OUT PUNICODE_STRING CompleteName,
+    IN OUT PUNICODE_STRING RemainingName,
+    IN OUT PVOID Context OPTIONAL,
+    IN PSECURITY_QUALITY_OF_SERVICE SecurityQos OPTIONAL,
+    OUT PVOID *Object
+)
+{
+    DPRINT("ParseProc() called\n");
+    *Object = NULL;
+    return STATUS_OBJECT_NAME_NOT_FOUND;
+}
+
+static
+BOOLEAN
+NTAPI
+OkayToCloseProc(
+    IN PEPROCESS Process OPTIONAL,
+    IN PVOID Object,
+    IN HANDLE Handle,
+    IN KPROCESSOR_MODE AccessMode
+)
+{
+    DPRINT("OkayToCloseProc() 0x%p, Handle 0x%p, AccessMask 0x%lX\n",
+           Object,
+           Handle,
+           AccessMode);
+    return TRUE;
+}
+
+static
+NTSTATUS
+NTAPI
+QueryNameProc(
+    IN PVOID Object,
+    IN BOOLEAN HasObjectName,
+    OUT POBJECT_NAME_INFORMATION ObjectNameInfo,
+    IN ULONG Length,
+    OUT PULONG ReturnLength,
+    IN KPROCESSOR_MODE AccessMode
+)
+{
+    DPRINT("QueryNameProc() 0x%p, HasObjectName %d, Len %lu, AccessMask 0x%lX\n",
+           Object,
+           HasObjectName,
+           Length,
+           AccessMode);
+
+    ObjectNameInfo = NULL;
+    ReturnLength = 0;
+    return STATUS_SUCCESS;
+}
+
+/*!
+ * @brief This function creates a test directory.
+ */
+static
+VOID
+ObInsert_CreateTestDirectory()
+{
+    NTSTATUS Status;
+
+    /* Initialize the directory name */
+    RtlInitUnicodeString(&DirectoryName, L"\\ObInsertTest");
+
+    /* Initialize object attributes for the directory */
+    InitializeObjectAttributes(&DummyObjectAttributes,
+                               &DirectoryName,
+                               OBJ_KERNEL_HANDLE | OBJ_PERMANENT | OBJ_CASE_INSENSITIVE,
+                               NULL,
+                               NULL);
+
+    /* Create the directory object */
+    Status = ZwCreateDirectoryObject(&DirectoryHandle,
+                                     DELETE,
+                                     &DummyObjectAttributes);
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+}
+
+/*!
+ * @brief This function creates dummy object type.
+ */
+static
+VOID
+ObInsert_CreateDummyType()
+{
+    NTSTATUS Status;
+
+    /* Initialize the dummy type name */
+    RtlInitUnicodeString(&DummyTypeName, L"ObInsertDummyType");
+
+    RtlZeroMemory(&ObjectTypeInitializer, sizeof(ObjectTypeInitializer));
+
+    ObjectTypeInitializer.PoolType = NonPagedPool;
+    ObjectTypeInitializer.Length = sizeof(ObjectTypeInitializer);
+    ObjectTypeInitializer.ValidAccessMask = OBJECT_TYPE_ALL_ACCESS;
+    ObjectTypeInitializer.MaintainHandleCount = TRUE;
+
+    /* Create the dummy object type (expected to fail initially) */
```

Comment:
Is it going to fail because it is missing the callback pointers?

Reply:
Added comments to this code in https://github.com/reactos/reactos/pull/6968/commits/38de371bdf839b8cdbefe958e4a61ef3a0c674e9
(Initially a failure is expected because `MaintainHandleCount` is `TRUE` and requires the open/close procedure to be non-zero)